hummingbot/market/kucoin/kucoin_order_book_tracker.py

```diff
@@ -0,0 +1,159 @@
+            self._refresh_tracking_loop()
+        )
+        self._order_book_diff_router_task = safe_ensure_future(
+            self._order_book_diff_router()
+        )
+        self._order_book_snapshot_router_task = safe_ensure_future(
+            self._order_book_snapshot_router()
+        )
+
+    async def _order_book_diff_router(self):
+        """
+        Route the real-time order book diff messages to the correct order book.
+        """
+        last_message_timestamp: float = time.time()
+        messages_queued: int = 0
+        messages_accepted: int = 0
+        messages_rejected: int = 0
+
+        while True:
+            try:
+                ob_message: OrderBookMessage = await self._order_book_diff_stream.get()
+                symbol: str = ob_message.symbol
+
+                if symbol not in self._tracking_message_queues:
+                    continue
+                message_queue: asyncio.Queue = self._tracking_message_queues[symbol]
+                # Check the order book's initial update ID. If it's larger, don't bother.
+                order_book: OrderBook = self._order_books[symbol]
+
+                if order_book.snapshot_uid > ob_message.update_id:
+                    messages_rejected += 1
+                    continue
+                await message_queue.put(ob_message)
+                messages_accepted += 1
+
+                # Log some statistics.
+                now: float = time.time()
+                if int(now / 60.0) > int(last_message_timestamp / 60.0):
+                    self.logger().debug("Diff messages processed: %d, rejected: %d, queued: %d",
+                                        messages_accepted,
+                                        messages_rejected,
+                                        messages_queued)
+                    messages_accepted = 0
+                    messages_rejected = 0
+                    messages_queued = 0
+
+                last_message_timestamp = now
+            except asyncio.CancelledError:
+                raise
+            except Exception:
+                self.logger().network(
+                    f"Unexpected error routing order book messages.",
+                    exc_info=True,
+                    app_warning_msg=f"Unexpected error routing order book messages. Retrying after 5 seconds."
+                )
+                await asyncio.sleep(5.0)
+
+    async def _track_single_book(self, symbol: str):
+        message_queue: asyncio.Queue = self._tracking_message_queues[symbol]
+        order_book: OrderBook = self._order_books[symbol]
+        last_message_timestamp: float = time.time()
+        diff_messages_accepted: int = 0
+
+        while True:
+            try:
+                message: OrderBookMessage = await message_queue.get()
+                if message.type is OrderBookMessageType.DIFF:
+                    # Kucoin websocket messages contain the entire order book state so they should be treated as snapshots
```

Comment:
Seems like Kucoin has snapshots with diffs. Please use this instead

https://docs.kucoin.com/#level-2-market-data

Reply:
I already implemented the level2-market data for in both restful and websocket for snapshot and diff respectively. Are you requesting that I remove the `elif message.type is OrderBookMessageType.SNAPSHOT:` statement?

Reply:
according to their docs diffs are updates. if you call apply_snapshot doesnt it replace the whole orderbook with just  orders in the diff?